### ISSUE TITLE
Check for a stack lock and wait until lock is removed to continue

### ISSUE THREAD

**Issue by bjaworski3** [pulumi/pulumi#9047]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

When running pulumi in CI/CD, sometimes multiple builds get triggered on the same stack and the builds fail due to the stack being locked. It would be really helpful if there was a command to check if a lock exists on the stack and then wait for a specified time for the lock to be cleared.

The biggest use case here is PRs that run a `pulumi preview` to view the requested changes. If there is a build already running on the stack the preview will fail. It would be great to have the PRs wait for the lock to clear and then run the preview after, to prevent failed builds just due to locks.

**Comment by Frassle** [comment]
Sounds like this could be done via a `--wait-for-lock` option on `preview`? Maybe with a seconds timeout so things don't get stuck forever. I don't think we can support cli args that optionally take a parameter, but I think the following could work:
```
pulumi preview --wait-for-lock=-1 # wait forever (default)
pulumi preview --wait-for-lock=0  # don't wait for lock
pulumi preview --wait-for-lock=60 # wait for a minute
```

Same option could be added to update and destroy.

**Comment by lukehoban** [comment]
This is also very related to https://github.com/pulumi/pulumi/issues/2073 (or possibly even ultimately the same issue).

> The biggest use case here is PRs that run a pulumi preview to view the requested changes. If there is a build already running on the stack the preview will fail. It would be great to have the PRs wait for the lock to clear and then run the preview after, to prevent failed builds just due to locks.

Note that previews don't take locks, so this particular case shouldn't be causing issues.

**Comment by Frassle** [comment]
> Note that previews don't take locks, so this particular case shouldn't be causing issues.

I think we had one version of the CLI go out which was taking locks on previews, but that's been reverted now.

**Comment by bjaworski3** [comment]
> > Note that previews don't take locks, so this particular case shouldn't be causing issues.
> 
> I think we had one version of the CLI go out which was taking locks on previews, but that's been reverted now.

Althought previews don't take locks now, they will fail if there is an apply currently in progress but you attempt to run a preview

**Comment by cody2094** [comment]
In my opinion, I see value in a queuing system that queues the deploys in the order they hit Pulumi. In a large team, having that would be extremely helpful in the CI/CD realm. I also feel like it makes sense in the skip-preview mode as well.

**Comment by terekete** [comment]
@Frassle - is there an option to wait for lock with the cli?

**Comment by Frassle** [comment]
Not currently, that's what this issue is tracking.

**Comment by nikhilro** [comment]
+1, got Pulumi up and running on Github Actions and ran into this immediately. Would love a `--wait-for-lock` flag for `pulumi up`

**Comment by kevinresol** [comment]
> In my opinion, I see value in a queuing system that queues the deploys in the order they hit Pulumi. In a large team, having that would be extremely helpful in the CI/CD realm. I also feel like it makes sense in the skip-preview mode as well.

+1 for a queue system instead of a simple wait

**Comment by lukehoban** [comment]
> +1 for a queue system instead of a simple wait

Just one note that this feature of queuing Pulumi operation is a core feature of [Pulumi Deployments](https://www.pulumi.com/product/pulumi-deployments/).  It's possible to do this queuing much more cleanly in the case of Pulumi Deployments because the compute needed to do the Pulumi operation can be started when it is dequeued.  

That said - definitely understand the desire for this feature for users who are using other CI systems - but in those cases the best we'll be able to do is likely to have the `pulumi` operation wait for the lock to be available.  When we add that, indeed a fair queue around this on the backend so that leases are granted in the order they were requested will be ideal, though quite likely will require the Cloud backend where we can have compute on the backend to manage the coordination with clients on granting and timing out clients correctly.

**Comment by aangelisc** [comment]
Could an alternative solution be a command that allows us to check for a lock? I expect this may be easier to implement and would allow users to make use of things like merge queues in GH for example.

**Comment by dancmeyers** [comment]
Seconding @aangelisc's comment.

I'm find with complex queuing logic etc being part of a commercial product like Pulumi Deployments, but we're using a different CI/CD solution to manage things apart from Pulumi as well, so we're unlikely to change that. And that component already _has_ queueing logic.

The biggest awkwardness for us is that the question "Does a lock currently exist for this stack?" is an `aws s3 ls ...` command (we're using an S3 backend, not Pulumi Cloud), where a later release of Pulumi might choose to change that lockfile location without it being very explicit/visible (because it 'should' be an internal implementation detail, not something end-users have to know).

We need to be able to check for the presence of a lock not for the case of queueing up deployments (that happens before `pulumi` is ever invoked, so there should only be one instance running at a time), but because an in-progress deployment which receives a `SIGKILL` (multiple ctrl-C locally, CI workflow cancellation, whatever) does not get a chance to clean up and remove the lockfile. So there isn't _actually_ a deployment running, but Pulumi thinks there is.

If there was just a `pulumi check-lock` or similar command which returned some message about the lock state and a 0 exit code if there was no lockfile, nonzero if there was, then it would be simple to use that to prompt for human involvement to clean up safely.

**Comment by nierob** [comment]
> If there was just a `pulumi check-lock` or similar command which returned some message about the lock state and a 0 exit code if there was no lockfile, nonzero if there was, then it would be simple to use that to prompt for human involvement to clean up safely.

I agree, it would be nice to have ability to control lock. For your case, we had similar problem. Now, we are detecting error message "Pulumi stack is locked, probably by another deployment", call `cancel` and re-try. It does not win beauty contest, but it works for a while now.
